### PR TITLE
Add audit flags to kube-apiserver

### LINKED ISSUE #1872: Add Audit flags to Kube-apiserver
https://kubernetes.io/docs/admin/audit

### PR DESCRIPTION
**Pull request by sethpollack**
fixes #1872

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/1879)

### CODE REVIEW

File: pkg/apis/kops/componentconfig.go [hunk 1]
```diff
@@ -420,6 +420,15 @@ type KubeAPIServerConfig struct {
 	// otherwise the host's root CA set will be used.
 	OIDCCAFile *string `json:"oidcCAFile,omitempty" flag:"oidc-ca-file"`
 
+	// If set, all requests coming to the apiserver will be logged to this file.
+	AuditLogPath *string `json:"auditLogPath,omitempty" flag:"audit-log-path"`
+	// The maximum number of days to retain old audit log files based on the timestamp encoded in their filename.
+	AuditLogMaxage *string `json:"auditLogMaxage,omitempty" flag:"audit-log-maxage"`
```
Comment (justinsb) [inline pkg/apis/kops/componentconfig.go]: So I think maxage, maxbackups and maxsize are all int (path is a string).  Could I ask you to change them to `int32`s?

```
staging/src/k8s.io/apiserver/pkg/server/options/audit.go:       fs.StringVar(&o.Path, "audit-log-path", o.Path,
staging/src/k8s.io/apiserver/pkg/server/options/audit.go:       fs.IntVar(&o.MaxAge, "audit-log-maxage", o.MaxBackups,
staging/src/k8s.io/apiserver/pkg/server/options/audit.go:       fs.IntVar(&o.MaxBackups, "audit-log-maxbackup", o.MaxBackups,
staging/src/k8s.io/apiserver/pkg/server/options/audit.go:       fs.IntVar(&o.MaxSize, "audit-log-maxsize", o.MaxSize,
```

File: pkg/apis/kops/componentconfig.go [hunk 2]
```diff
@@ -420,6 +420,15 @@ type KubeAPIServerConfig struct {
 	// otherwise the host's root CA set will be used.
 	OIDCCAFile *string `json:"oidcCAFile,omitempty" flag:"oidc-ca-file"`
 
+	// If set, all requests coming to the apiserver will be logged to this file.
+	AuditLogPath *string `json:"auditLogPath,omitempty" flag:"audit-log-path"`
```
Comment (chrislovecnm) [inline pkg/apis/kops/componentconfig.go]: Nice ... need this too

**Comment by chrislovecnm** [comment]
@k8s-bot ok to test

**Comment by justinsb** [comment]
A final question @sethpollack and this is your call (if you say you like it as-is, we'll go with that).

We have `auditLogMaxage`, `auditLogMaxbackup`, `auditLogMaxsize` which is the correct mapping to the flags.  But we don't _have_ to be the same as the flags.  Would you prefer `auditLogMaxAge`, `auditLogMaxBackup`, `auditLogMaxSize`.

And also consider `auditLogMaxBackups`?

Entirely your call - you're a real-world user of this feature, so whatever makes more sense to you...

**Comment by sethpollack** [comment]
Yeah, I can make that change.

**Comment by sethpollack** [comment]
like this?
```
	// If set, all requests coming to the apiserver will be logged to this file.
	AuditLogPath *string `json:"auditLogPath,omitempty" flag:"audit-log-path"`
	// The maximum number of days to retain old audit log files based on the timestamp encoded in their filename.
	AuditLogMaxAge *int32 `json:"auditLogMaxAge,omitempty" flag:"audit-log-maxage"`
	// The maximum number of old audit log files to retain.
	AuditLogMaxBackups *int32 `json:"auditLogMaxBackups,omitempty" flag:"audit-log-maxbackup"`
	// The maximum size in megabytes of the audit log file before it gets rotated. Defaults to 100MB.
	AuditLogMaxSize *int32 `json:"auditLogMaxSize,omitempty" flag:"audit-log-maxsize"`
```

**Comment by sethpollack** [comment]
I'm trying to re-run `make apimachinery` but I'm getting this error: `cannot find package "k8s.io/kubernetes/pkg/runtime/schema"`

**Comment by sethpollack** [comment]
@justinsb thanks!

**Comment by ottoyiu** [comment]
I tried out these flags, and set the AuditLogPath to be /var/log/kube-apiserver-audit.log. However, that file isn't being exposed as a host mount. Can anyone point me in the right direction in how to expose it just like /var/log/kube-apiserver.log?

**Comment by Grillz** [comment]
docs?

**Pull request merged**